helix-term/src/commands.rs

```diff
@@ -2083,10 +2083,23 @@ pub mod cmd {
         let future = doc.format_and_save(fmt);
         cx.jobs.add(Job::new(future).wait_before_exiting());
 
+        // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
+        // This might not be necessary since `path` is already coming in as a Cow<str> to begin with.
+        let path_text = path
+            .unwrap_or(&doc.relative_path().unwrap().to_string_lossy())
+            .to_string();
+        let (lines, columns) = (doc.text().len_lines(), doc.text().len_chars());
+
         if path.is_some() {
             let id = doc.id();
             let _ = cx.editor.refresh_language_server(id);
         }
+
+        cx.editor.set_status(format!(
+            "\"{}\" {}L, {}C written",
+            path_text, lines, columns,
+        ));
+
```

Comment:
As is, this could emit even if the write fails. I think you'll need to add a callback to the job that's being added in line 2084, and check in there if the result was Ok before emitting the message.

```rs
Job::new(future).with_callback().wait_before_exit()
```

Reply:
I'm having a lot of difficulty trying to get the information I need into my callback. None of the types I need to use implement `Send`, but I need to be able to send them across threads in order 

```rs
fn write_impl(cx: &mut compositor::Context, path: Option<&Cow<str>>) -> anyhow::Result<()> {
...
    // FIXME: Currently our message displays even if the program fails to write the data!
    // let future = doc.format_and_save(fmt);
    let callback = Box::pin(async move {
        let save_ok_fut = doc.format_and_save(fmt);
        let save_ok = save_ok_fut.await; // Placing `await` on the line above caused a different compilation error that is resolved by making another variable because lifetimes.
        let call: job::Callback = Box::new(move |&mut editor, &mut compositor| {
            // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
            // This might not be necessary since `path` is already coming in as a `Cow<str>`.
            let path_text = path
                .unwrap_or(&doc.relative_path().unwrap().to_string_lossy())
                .to_string();
            let (lines, columns) = (doc.text().len_lines(), doc.text().len_chars());
            cx.editor.set_status(format!(
                "\"{}\" {}L, {}C written",
                path_text, lines, columns,
            ));
        });

        Ok(call)
    });

    cx.jobs
        .add(Job::with_callback(callback).wait_before_exiting());

    // I'm sure using `path` down here would cause a different set of problems involving moving path into the callback, but I'm surprised I have yet to even get the callback working at all.
    if path.is_some() {
        let id = doc.id();
        let _ = cx.editor.refresh_language_server(id);
    }

    Ok(())
}
```

Attempting to build the code in its current state produces the following compilation errors:
```rs
   Compiling helix-term v0.6.0 (/home/.../helix/helix-term)
error[E0277]: `(dyn ClipboardProvider + 'static)` cannot be sent between threads safely
    --> helix-term/src/commands.rs:2089:39
     |
2089 |               let call: job::Callback = Box::new(move |&mut editor, &mut compositor| {
     |  _______________________________________^
2090 | |                 // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
2091 | |                 // This might not be necessary since `path` is already coming in as a `Cow<str>`.
2092 | |                 let path_text = path
...    |
2099 | |                 ));
2100 | |             });
     | |______________^ `(dyn ClipboardProvider + 'static)` cannot be sent between threads safely
     |
     = help: the trait `Send` is not implemented for `(dyn ClipboardProvider + 'static)`
     = note: required because of the requirements on the impl of `Send` for `Unique<(dyn ClipboardProvider + 'static)>`
     = note: required because it appears within the type `Box<(dyn ClipboardProvider + 'static)>`
     = note: required because it appears within the type `helix_view::Editor`
     = note: required because of the requirements on the impl of `Send` for `&mut helix_view::Editor`
note: required because it appears within the type `compositor::Context<'_>`
    --> helix-term/src/compositor.rs:30:12
     |
30   | pub struct Context<'a> {
     |            ^^^^^^^
     = note: required because of the requirements on the impl of `Send` for `&mut compositor::Context<'_>`
     = note: required because it appears within the type `[closure@helix-term/src/commands.rs:2089:48: 2100:14]`
     = note: required for the cast to the object type `dyn for<'r, 's> FnOnce(&'r mut helix_view::Editor, &'s mut compositor::Compositor) + Send`

error[E0277]: `(dyn for<'r> Fn(&'r mut helix_view::Editor) + 'static)` cannot be sent between threads safely
    --> helix-term/src/commands.rs:2089:39
     |
2089 |               let call: job::Callback = Box::new(move |&mut editor, &mut compositor| {
     |  _______________________________________^
2090 | |                 // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
2091 | |                 // This might not be necessary since `path` is already coming in as a `Cow<str>`.
2092 | |                 let path_text = path
...    |
2099 | |                 ));
2100 | |             });
     | |______________^ `(dyn for<'r> Fn(&'r mut helix_view::Editor) + 'static)` cannot be sent between threads safely
     |
     = help: the trait `Send` is not implemented for `(dyn for<'r> Fn(&'r mut helix_view::Editor) + 'static)`
     = note: required because of the requirements on the impl of `Send` for `Unique<(dyn for<'r> Fn(&'r mut helix_view::Editor) + 'static)>`
     = note: required because it appears within the type `Box<(dyn for<'r> Fn(&'r mut helix_view::Editor) + 'static)>`
     = note: required because it appears within the type `Motion`
     = note: required because it appears within the type `std::option::Option<Motion>`
     = note: required because it appears within the type `helix_view::Editor`
     = note: required because of the requirements on the impl of `Send` for `&mut helix_view::Editor`
note: required because it appears within the type `compositor::Context<'_>`
    --> helix-term/src/compositor.rs:30:12
     |
30   | pub struct Context<'a> {
     |            ^^^^^^^
     = note: required because of the requirements on the impl of `Send` for `&mut compositor::Context<'_>`
     = note: required because it appears within the type `[closure@helix-term/src/commands.rs:2089:48: 2100:14]`
     = note: required for the cast to the object type `dyn for<'r, 's> FnOnce(&'r mut helix_view::Editor, &'s mut compositor::Compositor) + Send`

error[E0277]: `std::cell::Cell<tendril::tendril::PackedUsize>` cannot be shared between threads safely
    --> helix-term/src/commands.rs:2089:39
     |
2089 |               let call: job::Callback = Box::new(move |&mut editor, &mut compositor| {
     |  _______________________________________^
2090 | |                 // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
2091 | |                 // This might not be necessary since `path` is already coming in as a `Cow<str>`.
2092 | |                 let path_text = path
...    |
2099 | |                 ));
2100 | |             });
     | |______________^ `std::cell::Cell<tendril::tendril::PackedUsize>` cannot be shared between threads safely
     |
     = help: within `tendril::tendril::NonAtomic`, the trait `Sync` is not implemented for `std::cell::Cell<tendril::tendril::PackedUsize>`
     = note: required because it appears within the type `tendril::tendril::NonAtomic`
     = note: required because of the requirements on the impl of `Send` for `tendril::tendril::Tendril<tendril::fmt::UTF8>`
     = note: required because it appears within the type `helix_core::Operation`
     = note: required because of the requirements on the impl of `Send` for `Unique<helix_core::Operation>`
     = note: required because it appears within the type `alloc::raw_vec::RawVec<helix_core::Operation>`
     = note: required because it appears within the type `Vec<helix_core::Operation>`
     = note: required because it appears within the type `ChangeSet`
     = note: required because it appears within the type `helix_view::Document`
     = note: required because of the requirements on the impl of `Send` for `&mut helix_view::Document`
     = note: required because it appears within the type `[closure@helix-term/src/commands.rs:2089:48: 2100:14]`
     = note: required for the cast to the object type `dyn for<'r, 's> FnOnce(&'r mut helix_view::Editor, &'s mut compositor::Compositor) + Send`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `helix-term` due to 3 previous errors
```

I've also tried creating a standalone `async fn` that takes everything I need as function arguments but I get tripped up by the same trait bounds.

```rs
// working title
async fn save_status_message(
    fmt_result: impl Future<Output = anyhow::Result<()>>,
    editor: &mut Editor,
    path: Option<&Cow<'_, str>>,
) -> anyhow::Result<Callback> {
    let doc = doc_mut!(editor);
    let save_ok_fut = fmt_result;
    let save_ok = save_ok_fut.await;
    // Same compilation errors occur because I need to hand everything off to this closure. Either that's a necessity or I need to find where all this information is inside the `Editor` and `Compositor` types.
    let call: job::Callback = Box::new(move |&mut editor, &mut compositor| { 
        // TODO?: Accept all valid unicode correctly (`to_string_lossy` is used to satisfy the type system)
        // This might not be necessary since `path` is already coming in as a `Cow<str>`.
        let path_text = path
            .unwrap_or(&doc.relative_path().unwrap().to_string_lossy())
            .to_string();
        let (lines, columns) = (doc.text().len_lines(), doc.text().len_chars());

        editor.set_status(format!(
            "\"{}\" {}L, {}C written",
            path_text, lines, columns,
        ));
    });

    Ok(call)
}
```
I've vaguely considered trying to implement this in `save_impl` in `document.rs,` however that module's callbacks don't take a reference to an `Editor`.

Am I just missing something painfully obvious or would implementing this feature correctly require reworking the API? Forgive me if I'm overcomplicating the problem, this is my first ever PR. All I want is a way to access the result of `format_and_save` so I can decide whether or not to print the message.

Reply:
You're trying to move the document in the closure, that won't work. You need to `let id = doc.id()` then use that id inside the closure to look up the document again.

Reply:
Okay, I see. I've implemented the changes accordingly. Thanks for your advice! Is there anything else that needs fixing?

Reply:
Bumping this to the top of the inbox :slightly_smiling_face: